### PR TITLE
ARROW-3803: [C++/Python] Merge C++ builds and tests, run Python tests in separate CI entries

### PR DESCRIPTION
**Pull request by wesm**
I found unfortunately that the conda-forge boost-cpp package is not fully compatible with Xcode 8.3, see https://issues.apache.org/jira/browse/ARROW-4056

We might have to build a vendored Boost in this CI entry to work around the problem (this is what the Ray project did when they also hit this issue)

### CODE REVIEW

**Comment by wesm** [comment]
We may be reaching the end of the road with Xcode 6.4. Travis is pulling support for the image in January 2019 according to messages in the build logs. 

Gandiva does not run properly (due to a strptime bug?) on 6.4 so we might move everything to 8.3. We should see what conda-forge is planning exactly, since if they are remaining on 6.4 then we may not be able to have Gandiva in the Python wheels

cc @pravindra @praveenbingo

**Comment by pravindra** [comment]
@wesm  current change lgtm i.e CI has gandiva with Xcode 8.3 but not Xcode 6.4.

If gandiva with Xcode 6.4 becomes important/required for any reason, we can spend some time on workarounds to make it work.

**Comment by wesm** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-4068 about Xcode 6.4. If you know what problems there are, could you post them there? I'm working on getting a reliable macOS box where I can do development and testing

**Comment by wesm** [comment]
I ran into https://issues.apache.org/jira/browse/ARROW-4070 here when trying to use ARROW_BOOST_VENDORED with ninja

File: .travis.yml [hunk 1]
```diff
@@ -136,6 +136,10 @@ matrix:
     - ARROW_TRAVIS_GANDIVA_JAVA=1
     - ARROW_TRAVIS_OPTIONAL_INSTALL=1
     - ARROW_BUILD_WARNING_LEVEL=CHECKIN
+    # ARROW-3803: The Xcode 8.3 image has Boost libraries in /usr/local/lib
+    # which can get loaded before the toolchain Boost libraries. May be
+    # possible to fix without static linking
```
Comment (fsaintjacques) [inline .travis.yml]: `LD_LIBRARY_PATH` is less invasive.

Comment (wesm) [reply .travis.yml]: LD_LIBRARY_PATH is already set... 

https://github.com/apache/arrow/blob/master/ci/travis_env_common.sh#L63

**Comment by wesm** [comment]
I think this is almost there. I'm just waiting on the Gandiva valgrind fixes to go in. If folks could take pity on me and try to not touch the CI scripts any more until this is merged that would be really great

**Comment by pravindra** [comment]
> I opened https://issues.apache.org/jira/browse/ARROW-4068 about Xcode 6.4. If you know what problems there are, could you post them there? I'm working on getting a reliable macOS box where I can do development and testing

will do. thanks.

**Comment by wesm** [comment]
Rebased. Will merge once the build runs

**Comment by wesm** [comment]
Running all the C++ unit tests under valgrind is taking about 23 minutes

```
Label Time Summary:
arrow-tests      = 369.38 sec*proc (46 tests)
gandiva-tests    = 575.75 sec*proc (27 tests)
parquet-tests    = 458.77 sec*proc (18 tests)
plasma-tests     =  14.58 sec*proc (2 tests)
unittest         = 1418.48 sec*proc (93 tests)
```

This is OK for now, but may not be scalable.

**Comment by wesm** [comment]
Overall this patch cuts about 10 minutes out of build times

build_times

Will merge as soon as Appveyor is looking ok

**Comment by wesm** [comment]
The Appveyor builds passed (Rust is still running -- I opened https://issues.apache.org/jira/browse/ARROW-4080)

**Pull request closed without merging**